File: ampere-kde/src/providers/anthropic.rs
use crate::creds::{self, ClaudeCredentials};
use crate::model::{Authority, Gauge, Snapshot, Unit};
use chrono::{DateTime, Utc};
use serde_json::Value;
use std::os::unix::fs::PermissionsExt;

const USAGE_URL: &str = "https://api.anthropic.com/api/oauth/usage";
const TOKEN_URL: &str = "https://platform.claude.com/v1/oauth/token";
const CLIENT_ID: &str = "9d1c250a-e61b-44d9-88ed-5944d1962f5e";
const OAUTH_BETA: &str = "oauth-2025-04-20";

pub fn fetch(client: &reqwest::blocking::Client) -> Snapshot {
    match load_and_fetch(client) {
        Ok(snapshot) => snapshot,
        Err(error) => unavailable(error),
    }
}

fn load_and_fetch(client: &reqwest::blocking::Client) -> Result<Snapshot, String> {
    let mut creds = creds::load_claude()?;

    if is_near_expiry(&creds) {
        if let Ok(fresh) = refresh(client, &creds.refresh_token) {
            creds = fresh;
        }
    }

    let (status, body) = get_usage(client, &creds.access_token)?;
    if status == 401 || status == 403 {
        let fresh = refresh(client, &creds.refresh_token)?;
        let (retry_status, retry_body) = get_usage(client, &fresh.access_token)?;
        if retry_status != 200 {
            return Err(format!("usage endpoint returned {retry_status} after refresh"));
        }
        return Ok(parse(&retry_body, &fresh));
    }
    if status != 200 {
        return Err(format!("usage endpoint returned {status}"));
    }
    Ok(parse(&body, &creds))
}

fn is_near_expiry(creds: &ClaudeCredentials) -> bool {
    creds.expires_at_ms > 0 && Utc::now().timestamp_millis() >= creds.expires_at_ms - 120_000
}

fn get_usage(client: &reqwest::blocking::Client, token: &str) -> Result<(u16, String), String> {
    let response = client
        .get(USAGE_URL)
        .header("authorization", format!("Bearer {token}"))
        .header("anthropic-beta", OAUTH_BETA)
        .header("accept", "application/json")
        .send()
        .map_err(|e| format!("request failed: {e}"))?;
    let status = response.status().as_u16();
    let body = response.text().unwrap_or_default();
    Ok((status, body))
}

fn parse(body: &str, creds: &ClaudeCredentials) -> Snapshot {
    let json: Value = serde_json::from_str(body).unwrap_or(Value::Null);
    let mut gauges = Vec::new();

    if let Some(limits) = json.get("limits").and_then(Value::as_array) {
        gauges.extend(limits.iter().filter_map(gauge_from_limit));
    }
    if gauges.is_empty() {
        gauges.extend(gauges_from_top_level(&json));
    }
    if let Some(extra) = gauge_from_extra_usage(&json) {
        gauges.push(extra);
    }

    Snapshot {
        provider_id: "anthropic".into(),
        provider_name: "Claude".into(),
        subtitle: subtitle(creds),
        authority: Authority::Live,
        source: "api.anthropic.com · live".into(),
        gauges,
        details: details(&json),
        note: None,
        error: None,
    }
}

fn details(json: &Value) -> Vec<(String, String)> {
    let mut details = Vec::new();
    if let Some(ts) = parse_ts(json.pointer("/five_hour/resets_at")) {
        details.push(("Session resets".into(), human_until(ts)));
    }
    if let Some(ts) = parse_ts(json.pointer("/seven_day/resets_at")) {
        details.push(("Weekly resets".into(), human_until(ts)));
    }
    let extra_enabled = json
        .pointer("/extra_usage/is_enabled")
        .and_then(Value::as_bool)
        .unwrap_or(false);
    details.push((
        "Extra usage credits".into(),
        if extra_enabled { "enabled".into() } else { "disabled".into() },
    ));
    details
}

fn human_until(ts: DateTime<Utc>) -> String {
    let seconds = (ts - Utc::now()).num_seconds();
    if seconds <= 0 {
        return "now".into();
    }
    let (days, hours, minutes) = (seconds / 86_400, (seconds % 86_400) / 3_600, (seconds % 3_600) / 60);
    if days > 0 {
        format!("in {days}d {hours}h")
    } else if hours > 0 {
        format!("in {hours}h {minutes}m")
    } else {
        format!("in {minutes}m")
    }
}

fn gauge_from_limit(item: &Value) -> Option<Gauge> {
    let kind = item.get("kind").and_then(Value::as_str)?;
    let percent = item.get("percent").and_then(Value::as_f64)?;
    let model = item
        .get("scope")
        .and_then(|s| s.get("model"))
        .and_then(|m| m.get("display_name"))
        .and_then(Value::as_str);
    let label = match (kind, model) {
        ("session", _) => "5-hour session".to_string(),
        ("weekly_all", _) => "Weekly · all models".to_string(),
        ("weekly_scoped", Some(name)) => format!("Weekly · {name}"),
        ("weekly_scoped", None) => "Weekly · scoped".to_string(),
        (other, Some(name)) => format!("{} · {name}", pretty(other)),
        (other, None) => pretty(other),
    };
    Some(Gauge {
        key: kind.to_string(),
        label,
        fraction: (percent / 100.0).clamp(0.0, 1.0),
        used: None,
        limit: None,
        unit: Unit::Percent,
        detail: None,
        resets_at: parse_ts(item.get("resets_at")),
        trusted_reset: kind == "session",
    })
}

fn gauges_from_top_level(json: &Value) -> Vec<Gauge> {
    [
        ("five_hour", "5-hour session", true),
        ("seven_day", "Weekly · all models", false),
    ]
    .into_iter()
    .filter_map(|(key, label, trusted)| {
        let obj = json.get(key)?;
        let utilization = obj.get("utilization").and_then(Value::as_f64)?;
        Some(Gauge {
            key: key.to_string(),
            label: label.to_string(),
            fraction: (utilization / 100.0).clamp(0.0, 1.0),
            used: None,
            limit: None,
            unit: Unit::Percent,
            detail: None,
            resets_at: parse_ts(obj.get("resets_at")),
            trusted_reset: trusted,
        })
    })
    .collect()
}

fn gauge_from_extra_usage(json: &Value) -> Option<Gauge> {
    let extra = json.get("extra_usage")?;
    if extra.get("is_enabled").and_then(Value::as_bool) != Some(true) {
        return None;
    }
    let utilization = extra.get("utilization").and_then(Value::as_f64)?;
    Some(Gauge {
        key: "extra_usage".into(),
        label: "Extra usage credits".into(),
        fraction: (utilization / 100.0).clamp(0.0, 1.0),
        used: extra.get("used_credits").and_then(Value::as_f64),
        limit: extra.get("monthly_limit").and_then(Value::as_f64),
        unit: Unit::Usd,
        detail: None,
        resets_at: None,
        trusted_reset: false,
    })
}

fn pretty(kind: &str) -> String {
    let mut chars = kind.replace('_', " ");
    if let Some(first) = chars.get_mut(0..1) {
        first.make_ascii_uppercase();
    }
    chars
}

fn subtitle(creds: &ClaudeCredentials) -> String {
    let plan = match creds.subscription_type.as_str() {
        "max" => "Max",
        "pro" => "Pro",
        other => other,
    };
    match creds
        .rate_limit_tier
        .rsplit('_')
        .next()
        .filter(|s| s.ends_with('x'))
    {
        Some(mult) => format!("{plan} · {}×", mult.trim_end_matches('x')),
        None => plan.to_string(),
    }
}

fn parse_ts(value: Option<&Value>) -> Option<DateTime<Utc>> {
    let raw = value?.as_str()?;
    DateTime::parse_from_rfc3339(raw)
        .ok()
        .map(|dt| dt.with_timezone(&Utc))
}

fn refresh(
    client: &reqwest::blocking::Client,
    refresh_token: &str,
) -> Result<ClaudeCredentials, String> {
    if refresh_token.is_empty() {
        return Err("no refresh token — run `claude` to sign in".into());
    }
    let payload = serde_json::json!({
        "grant_type": "refresh_token",
        "refresh_token": refresh_token,
        "client_id": CLIENT_ID,
    });
    let response = client
        .post(TOKEN_URL)
        .json(&payload)
        .send()
        .map_err(|e| format!("refresh request failed: {e}"))?;
    if !response.status().is_success() {
        return Err(format!("token refresh returned {}", response.status()));
    }
    let json: Value = response
        .json()
        .map_err(|e| format!("bad refresh response: {e}"))?;
    let access = json
        .get("access_token")
        .and_then(Value::as_str)
        .ok_or("no access_token in refresh response")?
        .to_string();
    let new_refresh = json
        .get("refresh_token")
        .and_then(Value::as_str)
        .unwrap_or(refresh_token)
        .to_string();
    let expires_in = json.get("expires_in").and_then(Value::as_i64).unwrap_or(28_800);
    let expires_at_ms = Utc::now().timestamp_millis() + expires_in * 1000;

    write_back(&access, &new_refresh, expires_at_ms)?;
    creds::load_claude()
}

/// Rotate the OAuth tokens back into the credentials file without disturbing
/// any sibling keys (e.g. `mcpOAuth`), writing atomically at mode 600.
fn write_back(access: &str, refresh: &str, expires_at_ms: i64) -> Result<(), String> {
    let path = creds::claude_credentials_path();
    let raw = std::fs::read_to_string(&path).map_err(|e| e.to_string())?;
    let mut json: Value = serde_json::from_str(&raw).map_err(|e| e.to_string())?;

    let oauth = json
        .get_mut("claudeAiOauth")
        .and_then(Value::as_object_mut)
        .ok_or("credentials file missing claudeAiOauth")?;
    oauth.insert("accessToken".into(), Value::String(access.to_string()));
    oauth.insert("refreshToken".into(), Value::String(refresh.to_string()));
    oauth.insert("expiresAt".into(), Value::Number(expires_at_ms.into()));

    let serialized = serde_json::to_vec_pretty(&json).map_err(|e| e.to_string())?;
    let tmp = path.with_extension("json.ampere-tmp");
    std::fs::write(&tmp, &serialized).map_err(|e| e.to_string())?;
    std::fs::set_permissions(&tmp, std::fs::Permissions::from_mode(0o600))
        .map_err(|e| e.to_string())?;
    std::fs::rename(&tmp, &path).map_err(|e| e.to_string())?;
    Ok(())
}

fn unavailable(error: String) -> Snapshot {
    Snapshot {
        provider_id: "anthropic".into(),
        provider_name: "Claude".into(),
        subtitle: "Claude Max".into(),
        authority: Authority::Unavailable,
        source: "api.anthropic.com · unreachable".into(),
        gauges: Vec::new(),
        details: Vec::new(),
        note: None,
        error: Some(error),
    }
}

File: tokenmaxxing-macos/Sources/Tokenmaxxing/Providers/Anthropic.swift
import Foundation

/// Live Claude Max/Pro quota from the same OAuth endpoint Claude Code's
/// `/usage` command uses. Survives HTTP 429 by serving the last good body from
/// disk and imposing a multi-minute cooldown so we stop hammering the endpoint.
enum AnthropicProvider {
    static let usageURL = URL(string: "https://api.anthropic.com/api/oauth/usage")!
    static let tokenURL = URL(string: "https://platform.claude.com/v1/oauth/token")!
    static let clientId = "9d1c250a-e61b-44d9-88ed-5944d1962f5e"
    static let oauthBeta = "oauth-2025-04-20"
    static let liveInterval: TimeInterval = 120
    static let rateLimitFloor: TimeInterval = 5 * 60
    static let failCooldown: TimeInterval = 90

    struct FetchResult {
        var snapshot: Snapshot
        var cooldown: TimeInterval
        var fresh: Bool
    }

    static func fetch() async -> FetchResult {
        do {
            let (body, creds) = try await loadAndFetch()
            saveUsageCache(body)
            return FetchResult(snapshot: parse(body, creds: creds), cooldown: liveInterval, fresh: true)
        } catch let err as RateLimitedError {
            let cooldown = max(err.retryAfter ?? rateLimitFloor, rateLimitFloor)
            return FetchResult(
                snapshot: fallbackSnapshot(message: err.message, reason: "rate limited"),
                cooldown: cooldown,
                fresh: false
            )
        } catch {
            return FetchResult(
                snapshot: fallbackSnapshot(message: "\(error)", reason: "cached"),
                cooldown: failCooldown,
                fresh: false
            )
        }
    }

    private struct RateLimitedError: Error {
        var message: String
        var retryAfter: TimeInterval?
    }

    private static func loadAndFetch() async throws -> ([String: Any], ClaudeCredentials) {
        var creds = try Creds.loadClaude()
        let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
        if creds.expiresAtMs > 0, nowMs >= creds.expiresAtMs - 120_000 {
            if let fresh = try? await refresh(creds.refreshToken) { creds = fresh }
        }

        let first = try await getUsage(creds.accessToken)
        if first.status == 401 || first.status == 403 {
            let fresh = try await refresh(creds.refreshToken)
            let retry = try await getUsage(fresh.accessToken)
            try throwIfBad(retry.status, retryAfter: retry.retryAfter)
            return (retry.json, fresh)
        }
        try throwIfBad(first.status, retryAfter: first.retryAfter)
        return (first.json, creds)
    }

    private static func throwIfBad(_ status: Int, retryAfter: TimeInterval?) throws {
        if status == 200 { return }
        if status == 429 {
            throw RateLimitedError(message: "usage endpoint rate limited (HTTP 429)", retryAfter: retryAfter)
        }
        throw QuotaError.message("usage endpoint returned \(status)")
    }

    private static func getUsage(_ token: String) async throws -> (status: Int, json: [String: Any], retryAfter: TimeInterval?) {
        var request = URLRequest(url: usageURL)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "authorization")
        request.setValue(oauthBeta, forHTTPHeaderField: "anthropic-beta")
        request.setValue("application/json", forHTTPHeaderField: "accept")
        request.setValue("tokenmaxxing/0.2.0 (+https://github.com/guitaripod/tokenmaxxing)", forHTTPHeaderField: "user-agent")
        let (data, response) = try await URLSession.shared.data(for: request)
        let http = response as? HTTPURLResponse
        let status = http?.statusCode ?? 0
        let retryAfter: TimeInterval? = {
            guard let raw = http?.value(forHTTPHeaderField: "Retry-After")?.trimmingCharacters(in: .whitespaces),
                  let seconds = TimeInterval(raw)
            else { return nil }
            return max(1, seconds)
        }()
        let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        return (status, json, retryAfter)
    }

    private static func fallbackSnapshot(message: String, reason: String) -> Snapshot {
        if let body = loadUsageCache(), let creds = try? Creds.loadClaude() {
            var snap = parse(body, creds: creds)
            snap.source = "api.anthropic.com · \(reason)"
            snap.note = "Showing last good reading — \(message)"
            snap.error = nil
            return snap
        }
        return unavailable(message)
    }

    private static var cacheURL: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appending(path: "Library/Application Support/tokenmaxxing/claude_usage_cache.json")
    }

    private static func saveUsageCache(_ body: [String: Any]) {
        let payload: [String: Any] = [
            "saved_at_ms": Int64(Date().timeIntervalSince1970 * 1000),
            "body": body,
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: payload, options: [.prettyPrinted]) else { return }
        let url = cacheURL
        try? FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try? data.write(to: url, options: [.atomic])
    }

    private static func loadUsageCache() -> [String: Any]? {
        guard let data = try? Data(contentsOf: cacheURL),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let body = root["body"] as? [String: Any]
        else { return nil }
        return body
    }

    private static func parse(_ json: [String: Any], creds: ClaudeCredentials) -> Snapshot {
        var gauges: [Gauge] = []
        if let limits = json["limits"] as? [[String: Any]] {
            gauges = limits.compactMap(gaugeFromLimit)
        }
        if gauges.isEmpty { gauges = gaugesFromTopLevel(json) }
        if let extra = gaugeFromExtra(json) { gauges.append(extra) }

        return Snapshot(
            providerId: "anthropic",
            providerName: "Claude",
            subtitle: subtitle(creds),
            authority: .live,
            source: "api.anthropic.com · live",
            gauges: gauges,
            details: details(json),
            note: nil,
            error: nil,
            spend: spendInfo(json)
        )
    }

    private static func gaugeFromLimit(_ item: [String: Any]) -> Gauge? {
        guard let kind = item["kind"] as? String,
              let percent = (item["percent"] as? NSNumber)?.doubleValue
        else { return nil }
        let model = ((item["scope"] as? [String: Any])?["model"] as? [String: Any])?["display_name"] as? String
        let label: String
        switch (kind, model) {
        case ("session", _): label = "5-hour session"
        case ("weekly_all", _): label = "Weekly · all models"
        case ("weekly_scoped", .some(let name)): label = "Weekly · \(name)"
        case ("weekly_scoped", .none): label = "Weekly · scoped"
        case (let other, .some(let name)): label = "\(pretty(other)) · \(name)"
        case (let other, .none): label = pretty(other)
        }
        return Gauge(
            key: kind, label: label,
            fraction: min(1, max(0, percent / 100)),
            used: nil, limit: nil, unit: .percent,
            detail: nil, resetsAt: parseTimestamp(item["resets_at"]),
            trustedReset: kind == "session",
            apiSeverity: severity(from: item["severity"] as? String),
            isActive: (item["is_active"] as? NSNumber)?.boolValue ?? false
        )
    }

    /// Prefer the server's own severity over the fraction-derived threshold.
    private static func severity(from value: String?) -> Severity? {
        switch value {
        case "critical": return .critical
        case "warn", "warning": return .warn
        case "normal", "ok": return .nominal
        default: return nil
        }
    }

    /// The prepaid / pay-as-you-go credit block, when the plan exposes one.
    private static func spendInfo(_ json: [String: Any]) -> SpendInfo? {
        guard let spend = json["spend"] as? [String: Any] else { return nil }
        return SpendInfo(
            enabled: (spend["enabled"] as? NSNumber)?.boolValue ?? false,
            used: money(spend["used"]) ?? 0,
            limit: money(spend["limit"]),
            balance: money(spend["balance"]),
            canPurchase: (spend["can_purchase_credits"] as? NSNumber)?.boolValue ?? false,
            disclaimer: (spend["disclaimer"] as? String).map(stripMarkdownLinks)
        )
    }

    /// Accepts a bare dollar number or a `{amount_minor, exponent}` object → dollars.
    private static func money(_ value: Any?) -> Double? {
        if let n = value as? NSNumber, !(value is NSNull) { return n.doubleValue }
        guard let dict = value as? [String: Any],
              let minor = (dict["amount_minor"] as? NSNumber)?.doubleValue
        else { return nil }
        let exponent = (dict["exponent"] as? NSNumber)?.intValue ?? 2
        return minor / pow(10.0, Double(exponent))
    }

    /// Reduce `[label](url)` markdown links to their label for plain rendering.
    private static func stripMarkdownLinks(_ text: String) -> String {
        var out = ""
        let chars = Array(text)
        var i = 0
        while i < chars.count {
            if chars[i] == "[" {
                i += 1
                var label = ""
                while i < chars.count, chars[i] != "]" { label.append(chars[i]); i += 1 }
                if i < chars.count { i += 1 } // skip ']'
                if i < chars.count, chars[i] == "(" {
                    while i < chars.count, chars[i] != ")" { i += 1 }
                    if i < chars.count { i += 1 } // skip ')'
                }
                out += label
            } else {
                out.append(chars[i])
                i += 1
            }
        }
        return out
    }

    private static func gaugesFromTopLevel(_ json: [String: Any]) -> [Gauge] {
        [("five_hour", "5-hour session", true), ("seven_day", "Weekly · all models", false)]
            .compactMap { key, label, trusted in
                guard let obj = json[key] as? [String: Any],
                      let utilization = (obj["utilization"] as? NSNumber)?.doubleValue
                else { return nil }
                return Gauge(
                    key: key, label: label,
                    fraction: min(1, max(0, utilization / 100)),
                    used: nil, limit: nil, unit: .percent,
                    detail: nil, resetsAt: parseTimestamp(obj["resets_at"]),
                    trustedReset: trusted
                )
            }
    }

    private static func gaugeFromExtra(_ json: [String: Any]) -> Gauge? {
        guard let extra = json["extra_usage"] as? [String: Any],
              (extra["is_enabled"] as? NSNumber)?.boolValue == true,
              let utilization = (extra["utilization"] as? NSNumber)?.doubleValue
        else { return nil }
        return Gauge(
            key: "extra_usage", label: "Extra usage credits",
            fraction: min(1, max(0, utilization / 100)),
            used: (extra["used_credits"] as? NSNumber)?.doubleValue,
            limit: (extra["monthly_limit"] as? NSNumber)?.doubleValue,
            unit: .usd, detail: nil, resetsAt: nil, trustedReset: false
        )
    }

    private static func details(_ json: [String: Any]) -> [Detail] {
        var details: [Detail] = []
        if let ts = parseTimestamp((json["five_hour"] as? [String: Any])?["resets_at"]) {
            details.append(Detail(key: "Session resets", value: "in \(Gauge.humanize(until: ts))"))
        }
        if let ts = parseTimestamp((json["seven_day"] as? [String: Any])?["resets_at"]) {
            details.append(Detail(key: "Weekly resets", value: "in \(Gauge.humanize(until: ts))"))
        }
        let extraEnabled = ((json["extra_usage"] as? [String: Any])?["is_enabled"] as? NSNumber)?.boolValue ?? false
        details.append(Detail(key: "Extra usage credits", value: extraEnabled ? "enabled" : "disabled"))
        return details
    }

    private static func subtitle(_ creds: ClaudeCredentials) -> String {
        let plan: String
        switch creds.subscriptionType {
        case "max": plan = "Max"
        case "pro": plan = "Pro"
        default: plan = creds.subscriptionType
        }
        if let multiplier = creds.rateLimitTier.split(separator: "_").last, multiplier.hasSuffix("x") {
            return "\(plan) · \(multiplier.dropLast())×"
        }
        return plan
    }

    private static func pretty(_ kind: String) -> String {
        kind.replacingOccurrences(of: "_", with: " ").capitalized
    }

    private static func parseTimestamp(_ value: Any?) -> Date? {
        guard let raw = value as? String else { return nil }
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = iso.date(from: raw) { return date }
        iso.formatOptions = [.withInternetDateTime]
        if let range = raw.range(of: #"\.\d+"#, options: .regularExpression) {
            var trimmed = raw
            trimmed.removeSubrange(range)
            return iso.date(from: trimmed)
        }
        return iso.date(from: raw)
    }

    private static func refresh(_ refreshToken: String) async throws -> ClaudeCredentials {
        guard !refreshToken.isEmpty else {
            throw QuotaError.message("no refresh token — run `claude` to sign in")
        }
        var request = URLRequest(url: tokenURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "grant_type": "refresh_token",
            "refresh_token": refreshToken,
            "client_id": clientId,
        ])
        let (data, response) = try await URLSession.shared.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            throw QuotaError.message("token refresh returned \(status)")
        }
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let access = json["access_token"] as? String
        else { throw QuotaError.message("no access_token in refresh response") }
        let newRefresh = json["refresh_token"] as? String ?? refreshToken
        let expiresIn = (json["expires_in"] as? NSNumber)?.int64Value ?? 28_800
        let expiresAtMs = Int64(Date().timeIntervalSince1970 * 1000) + expiresIn * 1000
        try Creds.writeBackClaude(accessToken: access, refreshToken: newRefresh, expiresAtMs: expiresAtMs)
        return try Creds.loadClaude()
    }

    private static func unavailable(_ error: String) -> Snapshot {
        Snapshot(
            providerId: "anthropic",
            providerName: "Claude",
            subtitle: "Claude Max",
            authority: .unavailable,
            source: "api.anthropic.com · unreachable",
            gauges: [],
            details: [],
            note: nil,
            error: error
        )
    }
}
